lib/core/src/connection_pool.cpp

```diff
@@ -301,27 +308,52 @@ namespace irods
 
             // Check if any resources have been modified.
             if (options_.refresh_connections_when_resource_changes_detected) {
-                for (auto&& row : irods::query{ctx.conn.get(), "select max(RESC_MODIFY_TIME)"}) {
+                // Capture whether the resources are in sync (or not). It's important that each
+                // resource property be updated to avoid unnecessary refreshes of the connection.
+                auto in_sync = true;
+
+                auto q =
+                    irods::experimental::query_builder{}
+                        .options(NO_DISTINCT | RETURN_TOTAL_ROW_COUNT) // NOLINT(hicpp-signed-bitwise)
+                        .build(*ctx.conn, "select order_desc(RESC_MODIFY_TIME), order_desc(RESC_MODIFY_TIME_MILLIS)");
+
+                if (const auto v = q.total_row_count(); v != ctx.resc_count) {
+                    in_sync = false;
+                    ctx.resc_count = v;
+                }
+
+                if (ctx.resc_count > 0) {
+                    const auto row = q.front();
+
                     try {
-                        if (const auto mtime = std::stoull(row[0]); mtime > ctx.latest_resc_mtime) {
-                            ctx.latest_resc_mtime = mtime;
-                            return false;
+                        auto latest_mtime = fmt::format("{}.{}", row[0], row[1]);
+
+                        if (latest_mtime != ctx.latest_resc_mtime) {
+                            in_sync = false;
+                            ctx.latest_resc_mtime = std::move(latest_mtime);
                         }
                     }
                     catch (...) {
                         // In theory, we should never reach this block. However, if we do, just
                         // go ahead and refresh the connection. At the very least, the agent will
                         // reflect the latest state in the catalog for a small penalty of TCP socket
                         // reconstruction.
-                        return false;
+                        in_sync = false;
                     }
                 }
-            }
 
-            // Check if the connection is still valid.
-            // This query will always succeed unless there's an issue in which case an exception will
-            // be thrown.
-            query{ctx.conn.get(), "select ZONE_NAME where ZONE_TYPE = 'local'"}; // NOLINT(bugprone-unused-raii)
+                // Only return false if it was determined that the connection needed to be synchronized
+                // with the catalog.
+                if (!in_sync) {
+                    return in_sync;
+                }
```

Comment:
This function returns `true` in the implicit else case here (see L362), so we could just return `in_sync` here in any case.

Reply:
Thinking about this a little more, if we change this to always return here, that will increase the chance of introducing a bug if a developer wants to add new refresh options after the resource change detection logic.

Conditionally returning here allows new refresh options to be added without risk of breaking the implementation.

Thoughts?

Reply:
either seems fine - but leave comments in the code to tell the next developer the intent of the design.

Reply:
I added a comment explaining the reason for leaving this in.